### PR TITLE
Discovery Service: rescan threadsafe

### PR DESCRIPTION
**Pull request by eliska-n**
Error was found:
`Non-thread-safe operation invoked on an event loop other than the current one`

The callback from kazoo has to stay on the "other" thread created in the proactor service. 

Přemek told me to use `asyncio.run_coroutine_threadsafe` to "return" the callback into the main thread.
However, I need to somehow take care of the resulting `Future` object as much as TaskService takes care of its tasks. 

It's a mess...

### CODE REVIEW

File: asab/api/discovery.py [hunk 1]
```diff
@@ -250,6 +259,20 @@ def get_data(item):
 			yield item, json.loads(item_data)
 
 
+	def _on_change(self, watched_event):
+		future = asyncio.run_coroutine_threadsafe(self._rescan_advertised_instances(), self.App.Loop)
```
Comment (PremyslCerny) [inline asab/api/discovery.py:263]: This future should only be run when the previous finished, i. e. self._concurrent_future.done() or self._concurrent_future is None.

File: asab/api/discovery.py [hunk 2]
```diff
@@ -28,17 +28,26 @@ def __init__(self, app, zkc, service_name="asab.DiscoveryService") -> None:
 		self._cache_lock = asyncio.Lock()
 		self._ready_event = asyncio.Event()
 
+		self._concurrent_future = None
```
Comment (PremyslCerny) [inline asab/api/discovery.py:31]: The original idea was to pass the future to the TaskService, if it is possible:

```
		future = asyncio.run_coroutine_threadsafe(self._rescan_advertised_instances(), self.App.Loop)
		self.App.TaskService.schedule([future])
```

In this case you would not need to care about the await of the task in finalize.

File: asab/api/discovery.py [hunk 3]
```diff
@@ -250,6 +259,20 @@ def get_data(item):
 			yield item, json.loads(item_data)
 
 
+	def _on_change(self, watched_event):
+		future = asyncio.run_coroutine_threadsafe(self._rescan_advertised_instances(), self.App.Loop)
+		if self._concurrent_future is None:
+			self._concurrent_future = future
+
+		elif not self._concurrent_future.done():
+			# if previous execution is still in progress, wait. it is blocking, but not in the main thread
```
Comment (eliska-n) [inline asab/api/discovery.py:268]: is it true? that it does not block the main thread? :pleading_face:

File: asab/api/discovery.py (hunk 2 above)
Comment (eliska-n) [reply asab/api/discovery.py:31]: :( to nefunguje

File: asab/api/discovery.py [hunk 4]
```diff
@@ -30,15 +30,22 @@ def __init__(self, app, zkc, service_name="asab.DiscoveryService") -> None:
 
 		self.App.PubSub.subscribe("Application.tick/300!", self._on_tick)
 		self.App.PubSub.subscribe("ZooKeeperContainer.state/CONNECTED!", self._on_zk_ready)
+		self.App.PubSub.subscribe("Discover.Rescan!", self._on_rescan)
```
[the remark this answers is not in the document]
Comment (ateska) [reply asab/api/discovery.py:33]: I agree

File: asab/api/discovery.py [hunk 5]
```diff
@@ -30,15 +30,22 @@ def __init__(self, app, zkc, service_name="asab.DiscoveryService") -> None:
...
 
 
 	def _on_tick(self, msg):
-		self._update_cache(msg)
+		self._update_cache()
 
+	async def _on_rescan(self, msg):
+		# async method to provide threadsafe callback
+		self._update_cache()
```
Comment (ateska) [inline asab/api/discovery.py:41]: Be careful with this - it is a method that is / should be called on the non-main thread - we should not introduce this kind of methods.

**Review by ateska — CHANGES_REQUESTED** [review]

File: asab/api/discovery.py [hunk 6]
```diff
@@ -250,6 +257,10 @@ def get_data(item):
 			yield item, json.loads(item_data)
 
 
+	def _on_change(self, watched_event):
+		self.App.PubSub.publish_threadsafe("Discover.Rescan!")
```
Comment (ateska) [inline asab/api/discovery.py:261]: Comment this method to indicated clearly and loudly that this callback is happening on non-main thread.

Try more direct approach:

```
def _on_change(self, watched_event)
    self.App.Loop.call_soon_threadsafe(self._update_cache)
```

File: asab/api/discovery.py [hunk 7]
```diff
@@ -250,6 +257,10 @@ def get_data(item):
 			yield item, json.loads(item_data)
 
 
+	def _on_change(self, watched_event):
```
Comment (ateska) [inline asab/api/discovery.py:260]: `_on_change_threadsafe(...)`

**Review by ateska — CHANGES_REQUESTED** [review]

**Comment by eliska-n** [comment]
now it is final version

**Review by ateska — APPROVED** [review]

**Pull request merged**